test/distributed/_composable/test_checkpoint.py

```diff
@@ -219,6 +221,112 @@ def forward(self, x):
+
+            def forward(self, x):
+                if self.raise_exp and self.fwd_count == 0:
+                    raise RuntimeError("foo")
+                if self.raise_exp and self.fwd_count == 1:
+                    raise RuntimeError("bar")
+                if self.change_shape_in_recomp and self.fwd_count == 1:
+                    x.relu_()
+                random_tensor = torch.randn(1, 2)
+                x = self.a(x + random_tensor)
+                self.fwd_count += 1
+                return x
+
+        m = MyModel(True, False)
+        m0, m1, m2, m3 = (deepcopy(m) for _ in range(4))
+
+        # composable checkpoint does not support use_reentrant=True
+        with self.assertRaises(NotImplementedError):
+            checkpoint(m, use_reentrant=True)
+
+        # check giving an unsupported kwarg
+        with self.assertRaises(ValueError):
+            checkpoint(m0, foo="bar")
+
+        handled_fwd_exp = False
+        handled_recomp_exp = False
+
+        @contextmanager
+        def fwd_ctx(mod: MyModel):
+            try:
+                mod.raise_exp = False
+                yield
+            finally:
+                nonlocal handled_fwd_exp
+                handled_fwd_exp = True
+                mod.raise_exp = True
+
+        @contextmanager
+        def recomp_ctx(mod: MyModel):
+            try:
+                mod.raise_exp = False
+                yield
+            finally:
+                nonlocal handled_recomp_exp
+                handled_recomp_exp = True
+                mod.raise_exp = True
+
+        # Test different context functions
+        x = torch.randn(1, 2, requires_grad=True)
+        checkpoint(
+            m1, context_fn=lambda: (partial(fwd_ctx, m1)(), partial(recomp_ctx, m1)())
+        )
+        m1(x.clone()).sum().backward()
+        self.assertEqual((handled_fwd_exp, handled_recomp_exp), (True, True))
+
+        checkpoint(m2, context_fn=lambda: (nullcontext(), partial(recomp_ctx, m2)()))
+        with self.assertRaisesRegex(RuntimeError, "foo"):
+            m2(x.clone())
+
+        handled_fwd_exp = False  # Reset flag
+        checkpoint(m3, context_fn=lambda: (partial(fwd_ctx, m3)(), nullcontext()))
+        with self.assertRaisesRegex(RuntimeError, "bar"):
+            m3(x.clone()).sum().backward()
+        self.assertEqual(handled_fwd_exp, True)
+
+        # Test determinism check failure
+        m4 = MyModel(False, True)
+        m5 = deepcopy(m4)
+        # Determinism check should not throw an error,
+        # but autograd should throw a RuntimeError
+        checkpoint(m4, determinism_check="none")
+        with self.assertRaises(RuntimeError):
+            m4(x.clone()).sum().backward()
+        # Determinism check should throw a CheckpointError
+        checkpoint(m5, determinism_check="default")
+        with self.assertRaises(CheckpointError):
+            m5(x.clone()).sum().backward()
+
+        # Test preserving random state
```

Comment:
Should we check that running with `preserve_rng_state=True` matches without `checkpoint`?

Reply:
Yeah makes sense!